web-ng/src/app/services/auth/auth.guard.ts

```diff
@@ -14,36 +14,55 @@
  * limitations under the License.
  */
 import { Injectable } from '@angular/core';
-import { Router } from '@angular/router';
+import {
+  ActivatedRouteSnapshot,
+  CanActivate,
+  Router,
+  RouterStateSnapshot,
+} from '@angular/router';
 import { AuthService } from './auth.service';
 import { Observable, of } from 'rxjs';
 import { catchError, map } from 'rxjs/operators';
 import { environment } from '../../../environments/environment';
+import { User } from '../../shared/models/user.model';
 
 @Injectable({
   providedIn: 'root',
 })
-export class AuthGuard {
-  private static readonly SIGN_IN_URL = '/signin';
-
+export class AuthGuard implements CanActivate {
   constructor(private router: Router, private authService: AuthService) {}
 
-  canActivate(): Observable<boolean> {
+  canActivate(
+    route: ActivatedRouteSnapshot,
+    state: RouterStateSnapshot
+  ): Observable<boolean> {
     return this.authService.getUser$().pipe(
-      map(response => {
-        if (environment.useEmulators) {
-          return true;
+      map(user => {
+        if (state.url === '/signin') {
+          return this.handleSignin(user);
         }
-        if (response.isAuthenticated) {
+        if (this.isAuthenticated(user)) {
```

Comment:
I think this block can be made more self-describing; ideally the code would read more like the rules it implements. 

In practice, this might look something like this:
```ts
  canActivate(
    route: ActivatedRouteSnapshot,
    state: RouterStateSnapshot
  ): Observable<boolean> {
    return this.authService.getUser$().pipe(
      map(user => this.canUserActivate(user, route)),
      catchError(() => of(false))
    ).do(ok => if (!ok) this.router.navigate([AuthService.SIGN_IN_URL]); );
  }

  canUserActivate(user, route): boolean {
   {
      // User can access all routes when running local emulator.
      if (environment.useEmulators) {
         return true;
      }
      // Special rule for sign in URL - only allow access if *not* authenticated.
      if (route.something === AuthService.SIGN_IN_URL) {
        return !user.isAuthenticated;
      }
      // All other routers require sign in.
      return user.isAuthenticated;
  }

Reply:
thank you 🥇